From 03451747936b1be42f524db2f6906277fcc5517e Mon Sep 17 00:00:00 2001
From: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
Date: Wed, 8 Mar 2023 04:42:11 -0800
Subject: [PATCH] ath11k: Remove ieee80211_sta usage while processing packets 
 in tx path

Control would be null when the packet comes through 802.3 fast xmit
path.
Check for null cases and then process the control structure

Signed-off-by: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>
Signed-off-by: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
---
 drivers/net/wireless/ath/ath11k/mac.c | 10 +++-------
 1 file changed, 3 insertions(+), 7 deletions(-)

--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -7424,7 +7424,6 @@ static void ath11k_mac_op_tx(struct ieee
 	struct ath11k_sta *arsta = NULL;
 	struct ieee80211_mgmt *mgmt;
 	u32 info_flags = info->flags;
-	struct ieee80211_sta *sta = control->sta;
 	bool is_prb_rsp;
 	u16 frm_type = 0;
 	u8 tid, *qos_ctl;
@@ -7517,8 +7516,7 @@ static void ath11k_mac_op_tx(struct ieee
 			/* sta specific config has high precedence over vif specific config
 			 * Hence overwrite sta specific config if it is present
 			 */
-			if (sta) {
-				arsta = (struct ath11k_sta *)sta->drv_priv;
+			if (arsta) {
 				if (arsta->tid_cfg[tid].noack == WMI_PEER_TID_CONFIG_NOACK)
 					noack = true;
 
@@ -7536,8 +7534,7 @@ static void ath11k_mac_op_tx(struct ieee
 	if (ar->ab->nss.enabled)
 		ret = ath11k_nss_tx(arvif, skb);
 	else if (info_flags & IEEE80211_TX_CTL_HW_80211_ENCAP)
-		ret = ath11k_dp_tx_simple(ar, arvif, skb,
-					  (control->sta) ? control->sta->drv_priv : NULL);
+		ret = ath11k_dp_tx_simple(ar, arvif, skb, arsta);
 	else
 		ret = ath11k_dp_tx(ar, arvif, arsta, skb);
 
@@ -7548,13 +7545,10 @@ static void ath11k_mac_op_tx(struct ieee
 		return;
 	}
 
-	if (ath11k_debugfs_is_extd_tx_stats_enabled(ar) && sta) {
-		arsta = (struct ath11k_sta *)sta->drv_priv;
-		if (arsta) {
-			atomic_inc(&arsta->drv_tx_pkts.pkts_in);
-			if (!ret)
-				atomic_inc(&arsta->drv_tx_pkts.pkts_out);
-		}
+	if (ath11k_debugfs_is_extd_tx_stats_enabled(ar) && arsta) {
+		atomic_inc(&arsta->drv_tx_pkts.pkts_in);
+		if (!ret)
+			atomic_inc(&arsta->drv_tx_pkts.pkts_out);
 	}
 }
 
